### PR TITLE
💻  Fetch live clubs and events with React Query

### PR DESCRIPTION
**Pull request by jeffzh4ng**
## Screenshots
image

image

There are no images because our API doesn't have real images right [now](https://waterpark-staging.herokuapp.com/api/v1/clubs).

## Purpose
Fetches live clubs and events with React Query.

## Approach
The PR installs React Query as our frontend's http client. Check out the spike in Notion for more details on why React Query was chosen over SWR.

Creates a new folder until utils called `serde` (short for *ser*ialize and *de*seerialize). Another potential name for this folder can simply be `api`. This folder contains deserializers that convert DTO's to client-side types.

### CODE REVIEW

File: client/context/Base.tsx [hunk 1]
```diff
@@ -13,12 +16,14 @@ export interface Club {
...
-  instagramLink?: string
-  websiteLink?: string
+  links: {
+    bannerImage: string
+    iconImage: string
+    facebook?: string
+    twitter?: string
+    instagram?: string
+    website?: string
+  }
```
Comment (jeffzh4ng) [inline client/context/Base.tsx:26]: Changed frontend type to mirror backend type.

File: docker-compose.yml [hunk 2]
```diff
@@ -1,4 +1,4 @@
-version: "3.8"
```
Comment (xujustinj) [inline docker-compose.yml:1]: I feel like this isn't the first time we've seen these quotation marks change. Do we have conflicting Prettier setups somewhere?

File: client/utils/serde/clubs-and-events.ts [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  eventId: number,
+  club: ClubDTO
+): Event => {
+  const localZone = Intl.DateTimeFormat().resolvedOptions().timeZone
+
+  return {
+    ...serializedEvent,
+    id: eventId,
+    startTime: moment.tz('2022-11-18 10:55', 'Asia/Tokyo').local().tz(localZone),
+    endTime: moment.tz('2022-11-18 10:55', 'Asia/Tokyo').local().tz(localZone),
```
Comment (xujustinj) [inline client/utils/serde/clubs-and-events.ts:76]: Still hard-coding for now? I guess a future todo is to deserialize the actual date.

File: client/utils/serde/clubs-and-events.ts [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+  let eventId = 0
+  const deserializedEvents: Array<Event> = []
+
+  const deserializedClubs = serializedClubs.map((club, clubId) => {
+    return {
+      ...club,
+      id: clubId,
+      events: club.events.map((event) => {
+        const deserializedEvent = deserializeEvent(event, eventId++, club)
```
Comment (xujustinj) [inline client/utils/serde/clubs-and-events.ts:51]: so `eventId` seems to always be equal to `deserializedEvents.length`, maybe we can just use that?
```suggestion
        const deserializedEvent = deserializeEvent(event, deserializedEvents.length, club)
```

**Review by xujustinj — CHANGES_REQUESTED** [review]
Need to fix the `client/package-lock` again, sorry.

I'm not seeing any code that uses React Query to fetch from the backend, did I miss something?

File: client/components/club-details/ClubInfo.tsx [hunk 5]
```diff
@@ -63,15 +63,15 @@ interface ClubInfoProps {
 export const ClubInfo = ({ club }: ClubInfoProps) => {
   const router = useRouter()
 
-  const { name, description, iconURL, tags } = club
-  const links = [club.facebookLink, club.twitterLink, club.instagramLink, club.websiteLink]
-  const linksThatExist = links.filter((l) => l !== undefined)
+  const { name, description, links, tags } = club
+  const potentialLinks = [links.facebook, links.twitter, links.instagram, links.website]
+  const linksThatExist = potentialLinks.filter((l) => l !== undefined)
```
[the remark this answers is not in the document]
Comment (jeffzh4ng) [reply client/components/club-details/ClubInfo.tsx:68]: Changed.

File: docker-compose.yml (hunk 2 above)
Comment (jeffzh4ng) [reply docker-compose.yml:1]: Someone must be changing them to double, but we should be using single quotes.

File: client/utils/serde/clubs-and-events.ts (hunk 4 above)
Comment (jeffzh4ng) [reply client/utils/serde/clubs-and-events.ts:51]: you're a smart cs mans justin

File: client/utils/serde/clubs-and-events.ts (hunk 3 above)
Comment (jeffzh4ng) [reply client/utils/serde/clubs-and-events.ts]: woops, fixed. dates are working correctly :D

File: client/utils/serde/clubs-and-events.ts [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+  eventId: number,
+  club: ClubDTO
+): Event => {
+  const localZone = Intl.DateTimeFormat().resolvedOptions().timeZone
+
+  return {
+    ...serializedEvent,
+    id: eventId,
+    startTime: moment.tz(serializedEvent.startTime, 'Asia/Tokyo').local().tz(localZone),
+    endTime: moment.tz(serializedEvent.endTime, 'Asia/Tokyo').local().tz(localZone),
```
Comment (xujustinj) [inline client/utils/serde/clubs-and-events.ts:75]: I have no idea how moment timezones work (timezone documentation is unreadable to me), but I think we should either pass no timezone at all or GMT

**Review by xujustinj — APPROVED** [review]

File: client/utils/serde/clubs-and-events.ts (hunk 6 above)
Comment (ShreyasPrasad) [reply client/utils/serde/clubs-and-events.ts:75]: Yeh, I think that was solely for testing.

**Review by ShreyasPrasad — APPROVED** [review]

File: client/utils/serde/clubs-and-events.ts (hunk 6 above)
Comment (jeffzh4ng) [reply client/utils/serde/clubs-and-events.ts:75]: This timezone string (data) should actually come from the backend. I can change this client code to 'America/Toronto' on the client and we assume every time from the backend is EST.

Let's merge this in today before Spring BOT and then have a discussion later on.

**Pull request merged**